docs/spec/relayer/Packets.md

```diff
@@ -39,212 +18,67 @@ type SendPacketEvent {
-        return channel, proof
-    } else { return (nil, nil) }
-}
-```
-If LATEST_HEIGHT is passed as a parameter, the data should be read (and the corresponding proof created) 
-at the most recent height. 
-
-
-## Computing destination chain
-
-```golang
-func GetDestinationInfo(ev IBCEvent, chainA Chain) Chain {
-    switch ev.type {
-        case SendPacketEvent: 
-            channel, proof = GetChannel(chain, ev.sourcePort, ev.sourceChannel, ev.Height)
-            if proof == nil return nil
-                
-            connectionId = channel.connectionHops[0]
-            connection, proof = GetConnection(chain, connectionId, ev.Height) 
-            if proof == nil return nil
-                
-            clientState = GetClientState(chain, connection.clientIdentifier, ev.Height) 
-            return getHostInfo(clientState.chainID) 
-        ...    
-    }
-}
-```
-
 ## Datagram creation logic
 
 ### PacketRecv datagram creation
 
 ```golang
-func createPacketRecvDatagram(ev SendPacketEvent, chainA Chain, chainB Chain, installedHeight Height) PacketRecv {        
+func CreateDatagram(ev SendPacketEvent, 
+                    chainA Chain, 
+                    chainB Chain, 
+                    proofHeight Height) (PacketRecv, Error) {        
     
     // Stage 1 
     // Verify if packet is committed to chain A and it is still pending (commitment exists)
     
-    proofHeight = installedHeight - 1
     packetCommitment, packetCommitmentProof = 
         GetPacketCommitment(chainA, ev.sourcePort, ev.sourceChannel, ev.sequence, proofHeight)     
-    if packetCommitmentProof != nil { return nil }
+    if packetCommitmentProof == nil { return (nil, Error.RETRY) }
         
-    if packetCommitment == null OR
-       packetCommitment != hash(concat(ev.data, ev.timeoutHeight, ev.timeoutTimestamp)) { return nil }
+    if packetCommitment == nil OR
+       packetCommitment != hash(concat(ev.data, ev.timeoutHeight, ev.timeoutTimestamp)) { 
+            // invalid event; replace provider
+            ReplaceProvider(chainA)
+            return (nil, Error.DROP) 
+    }
             
     // Stage 2 
     // Execute checks IBC handler on chainB will execute
     
     channel, proof = GetChannel(chainB, ev.destPort, ev.destChannel, LATEST_HEIGHT)
-    if proof != nil { return nil }
+    if proof == nil { return (nil, Error.RETRY) }
     
-    if channel == null OR
-       channel.state != OPEN OR
-       ev.sourcePort != channel.counterpartyPortIdentifier OR
-       ev.sourceChannel != channel.counterpartyChannelIdentifier { return nil }
+    if channel != nil AND
+       (channel.state == CLOSED OR
+        ev.sourcePort != channel.counterpartyPortIdentifier OR
+        ev.sourceChannel != channel.counterpartyChannelIdentifier) { (nil, Error.DROP) } 
     
+    if channel == nil OR channel.state != OPEN  { (nil, Error.RETRY) } 
+    // TODO: Maybe we shouldn't even enter handle loop for packets if the corresponding channel is not open!
```

Comment:
Yes but I think here is the first time when we can make that determination.

Reply:
What I meant here is basically having supervisor logic that would decide when to call handleEvent based on event type, i.e., there would be some preconditions that are based on relayer state. For example, packet related events are preconditioned on channel and connections being open, all events are preconditioned on light client correctly installed, etc.

Reply:
Does this assume that the relayer keeps state for channel and connections? And the supervisor does not perform any queries to determine that. Trying to save on the number of queries :)

Reply:
Exactly. Supervisor will need anyway to keep some state (for example for pending packets so it can trigger timeouts), so why not keeping status of connections and channels, current block height and time, etc. It can save queries and potentially operates faster.